In [1]:
import pandas as pd
#import numpy as np
from sqlalchemy import create_engine
import warnings
import seaborn as sns

import matplotlib.pyplot as plt

%matplotlib inline

warnings.filterwarnings('ignore')

In [ ]:
#!pip install numpy==1.19.3

#As a workaround, you temporarily pin to NumPy 1.19.3 (`pip install numpy==1.19.3`) which has a fixed version of 
#OpenBLAS that works correctly in Windows 2004/20H2. 
#This fix broken other things (Linux, Windows-on-docker) and so could not be retained in NumPy 1.19.4.

In [2]:
import numpy as np

In [3]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'useducation'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

df = pd.read_sql_query('select * from useducation',con=engine)

# No need for an open connection, 
# because you're only doing a single query
engine.dispose()

In [4]:
df.head()

PRIMARY_KEY       STATE  YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0     1992_ALABAMA     ALABAMA  1992     NaN      2678885.0         304177.0   
1      1992_ALASKA      ALASKA  1992     NaN      1049591.0         106780.0   
2     1992_ARIZONA     ARIZONA  1992     NaN      3258079.0         297888.0   
3    1992_ARKANSAS    ARKANSAS  1992     NaN      1711959.0         178571.0   
4  1992_CALIFORNIA  CALIFORNIA  1992     NaN     26260025.0        2072470.0   

   STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0      1659028.0       715680.0          2653798.0                1481703.0   
1       720711.0       222100.0           972488.0                 498362.0   
2      1369815.0      1590376.0          3401580.0                1435908.0   
3       958785.0       574603.0          1743022.0                 964323.0   
4     16546514.0      7641041.0         27138832.0               14358922.0   

          ...           GRADES_4_G  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  \
0         ...              57948.0     58025.0      41167.0      471564.0   
1         ...               9748.0      8789.0       6714.0       79117.0   
2         ...              55433.0     49081.0      37410.0      437127.0   
3         ...              34632.0     36011.0      27651.0      281338.0   
4         ...             418418.0    363296.0     270675.0     3286034.0   

   GRADES_9_12_G  GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  \
0       196386.0      676174.0        208.327876        252.187522   
1        30847.0      112335.0               NaN               NaN   
2       175210.0      614881.0        215.253932        265.366278   
3       123113.0      405259.0        210.206028        256.312090   
4      1372011.0     4717112.0        208.398961        260.892247   

   AVG_READING_4_SCORE  AVG_READING_8_SCORE  
0           207.963517                  NaN  
1                  NaN           258.859712  
2           206.212716           262.169895  
3           208.634458           264.619665  
4           196.764414                  NaN  

[5 rows x 25 columns]

1. Determine all the variable types and find the fraction of the missing values for each variable.
2. Notice that the data has a time dimension (year). For this assignment, forget about time and treat all the observations as if they're from the same year. Choose a strategy to deal with the missing values for each variable. For which variables would filling in the missing values with some value make sense? For which might tossing out the records entirely make sense?
3. Now, take into account the time factor. Replicate your second answer, but this time, fill in the missing values by using a statistic that is calculated within the year of the observation. For example, if you want to fill in a missing value for a variable with the mean of that variable, calculate the mean by using only the observations for that specific year.
4. This time, fill in the missing values using interpolation (extrapolation).
5. Compare your results for the second, third, and fourth questions. Do you find any meaningful differences?

In [5]:
# data types are 2 objects, 1 intergers and 22 floats
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1492 entries, 0 to 1491
Data columns (total 25 columns):
PRIMARY_KEY                     1492 non-null object
STATE                           1492 non-null object
YEAR                            1492 non-null int64
ENROLL                          1229 non-null float64
TOTAL_REVENUE                   1280 non-null float64
FEDERAL_REVENUE                 1280 non-null float64
STATE_REVENUE                   1280 non-null float64
LOCAL_REVENUE                   1280 non-null float64
TOTAL_EXPENDITURE               1280 non-null float64
INSTRUCTION_EXPENDITURE         1280 non-null float64
SUPPORT_SERVICES_EXPENDITURE    1280 non-null float64
OTHER_EXPENDITURE               1229 non-null float64
CAPITAL_OUTLAY_EXPENDITURE      1280 non-null float64
GRADES_PK_G                     1319 non-null float64
GRADES_KG_G                     1360 non-null float64
GRADES_4_G                      1361 non-null float64
GRADES_8_G                      1

In [6]:
df.isnull()

PRIMARY_KEY  STATE   YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0           False  False  False    True          False            False   
1           False  False  False    True          False            False   
2           False  False  False    True          False            False   
3           False  False  False    True          False            False   
4           False  False  False    True          False            False   
5           False  False  False    True          False            False   
6           False  False  False    True          False            False   
7           False  False  False    True          False            False   
8           False  False  False    True          False            False   
9           False  False  False    True          False            False   
10          False  False  False    True          False            False   
11          False  False  False    True          False            False   
12          False  False  False    True          False            False   
13          False  False  False    True          False            False   
14          False  False  False    True          False            False   
15          False  False  False    True          False            False   
16          False  False  False    True          False            False   
17          False  False  False    True          False            False   
18          False  False  False    True          False            False   
19          False  False  False    True          False            False   
20          False  False  False    True          False            False   
21          False  False  False    True          False            False   
22          False  False  False    True          False            False   
23          False  False  False    True          False            False   
24          False  False  False    True          False            False   
25          False  False  False    True          False            False   
26          False  False  False    True          False            False   
27          False  False  False    True          False            False   
28          False  False  False    True          False            False   
29          False  False  False    True          False            False   
...           ...    ...    ...     ...            ...              ...   
1462        False  False  False    True           True             True   
1463        False  False  False    True           True             True   
1464        False  False  False    True           True             True   
1465        False  False  False    True           True             True   
1466        False  False  False    True           True             True   
1467        False  False  False    True           True             True   
1468        False  False  False    True           True             True   
1469        False  False  False    True           True             True   
1470        False  False  False    True           True             True   
1471        False  False  False    True           True             True   
1472        False  False  False    True           True             True   
1473        False  False  False    True           True             True   
1474        False  False  False    True           True             True   
1475        False  False  False    True           True             True   
1476        False  False  False    True           True             True   
1477        False  False  False    True           True             True   
1478        False  False  False    True           True             True   
1479        False  False  False    True           True             True   
1480        False  False  False    True           True             True   
1481        False  False  False    True           True             True   
1482        False  False  False    True           True             True   
1483        False  False  False

In [7]:
df.describe()

YEAR        ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
count  1492.000000  1.229000e+03   1.280000e+03     1.280000e+03   
mean   2004.433646  9.159308e+05   9.092082e+06     7.663723e+05   
std       7.393983  1.065280e+06   1.174519e+07     1.145242e+06   
min    1992.000000  4.386600e+04   4.656500e+05     3.102000e+04   
25%    1998.000000  2.583140e+05   2.186305e+06     1.893540e+05   
50%    2004.000000  6.483130e+05   5.079546e+06     4.033765e+05   
75%    2011.000000  1.014528e+06   1.085985e+07     8.289660e+05   
max    2017.000000  6.307022e+06   8.921726e+07     9.990221e+06   

       STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
count   1.280000e+03   1.280000e+03       1.280000e+03   
mean    4.216553e+06   4.109157e+06       9.196681e+06   
std     5.543072e+06   5.482971e+06       1.197813e+07   
min     0.000000e+00   2.209300e+04       4.816650e+05   
25%     1.153097e+06   7.158345e+05       2.165404e+06   
50%     2.537074e+06   2.055780e+06       5.234506e+06   
75%     5.080939e+06   4.768680e+06       1.074519e+07   
max     5.090457e+07   3.610526e+07       8.532013e+07   

       INSTRUCTION_EXPENDITURE  SUPPORT_SERVICES_EXPENDITURE  \
count             1.280000e+03                  1.280000e+03   
mean              4.762966e+06                  2.680331e+06   
std               6.293004e+06                  3.353349e+06   
min               2.655490e+05                  1.399630e+05   
25%               1.168032e+06                  6.357900e+05   
50%               2.657452e+06                  1.525406e+06   
75%               5.568028e+06                  3.229651e+06   
max               4.396452e+07                  2.605802e+07   

       OTHER_EXPENDITURE         ...              GRADES_4_G     GRADES_8_G  \
count       1.229000e+03         ...             1361.000000    1361.000000   
mean        4.292046e+05         ...            64538.927259   64271.057311   
std         5.340693e+05         ...            80293.387625   78910.545776   
min         1.154100e+04         ...              633.000000     437.000000   
25%         1.028310e+05         ...            13739.000000   13342.000000   
50%         2.715960e+05         ...            43272.000000   43339.000000   
75%         5.186000e+05         ...            75481.000000   76566.000000   
max         3.995951e+06         ...           493415.000000  500143.000000   

        GRADES_12_G  GRADES_1_8_G  GRADES_9_12_G  GRADES_ALL_G  \
count    1361.00000  1.361000e+03   1.361000e+03  1.319000e+03   
mean    54268.92432  5.192140e+05   2.470714e+05  8.024415e+05   
std     67889.27623  6.436384e+05   3.077056e+05  9.707031e+05   
min       311.00000  4.878000e+03   1.808000e+03  7.254000e+03   
25%     10919.00000  1.106260e+05   5.147100e+04  1.810650e+05   
50%     36735.00000  3.457750e+05   1.642600e+05  5.503420e+05   
75%     67460.00000  6.119050e+05   2.905020e+05  9.282755e+05   
max    498403.00000  3.929869e+06   2.013687e+06  5.944746e+06   

       AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
count        536.000000        532.000000           533.000000   
mean         234.768293        278.414711           218.866154   
std           10.221511         10.159722             7.769616   
min          187.134670        232.831510           178.557612   
25%          229.694352        272.761598           214.663401   
50%          237.238552        280.618803           220.416034   
75%          241.995486        285.347428           223.999337   
max          253.420961        300.568235           236.773867   

       AVG_READING_8_SCORE  
count           498.000000  
mean            263.661132  
std               6.803725  
min             236.379102  
25%             259.533171  
50%             265.010912  
75%             268.190121  
max             280.499130  

[8 rows x 23 columns]

In [8]:
# finding percentages of missing values
df.isnull().sum()*100/df.isnull().count()

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                          17.627346
TOTAL_REVENUE                   14.209115
FEDERAL_REVENUE                 14.209115
STATE_REVENUE                   14.209115
LOCAL_REVENUE                   14.209115
TOTAL_EXPENDITURE               14.209115
INSTRUCTION_EXPENDITURE         14.209115
SUPPORT_SERVICES_EXPENDITURE    14.209115
OTHER_EXPENDITURE               17.627346
CAPITAL_OUTLAY_EXPENDITURE      14.209115
GRADES_PK_G                     11.595174
GRADES_KG_G                      8.847185
GRADES_4_G                       8.780161
GRADES_8_G                       8.780161
GRADES_12_G                      8.780161
GRADES_1_8_G                     8.780161
GRADES_9_12_G                    8.780161
GRADES_ALL_G                    11.595174
AVG_MATH_4_SCORE                64.075067
AVG_MATH_8_SCORE                64.343164
AVG_READING_4_SCORE             64

2. Notice that the data has a time dimension (year). For this assignment, forget about time and treat all the observations as if they're from the same year. Choose a strategy to deal with the missing values for each variable. For which variables would filling in the missing values with some value make sense? For which might tossing out the records entirely make sense?

ANSWER: The missing data is continuous and mostly in enrollment, the revenue columns and the average columns. If all are the same year, the we can impute the missing enrollments with the mean for the entire dataset. For the revenue and expenditure vars, it makes sense to impute missing values for each with the median since we're looking at money grouped by state because each is missing the same percentage of data, the percentage missing is somewhat small and we don't skew the data especially if we were analyzing revenue against expenditure. For the GRADES variable, the percent missing is even smaller. GRADES would likely be an important variable to analyze and we only have 1492 so we want to keep them and impute a mean by state.  

The remaining AVG variables have the highest percent missing and they are only missing every other grouping. I would impute the missing values by forward filling. 

In [9]:
df.YEAR.unique()

array([1992, 2001, 1993, 2016, 1994, 1995, 1996, 1997, 1998, 1999, 2017,
       2000, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011,
       2012, 2013, 2014, 2015], dtype=int64)

In [10]:
df.YEAR.value_counts()

2004    59
2008    59
1999    59
2000    59
2001    59
2002    59
2003    59
2005    59
2007    59
2006    59
2009    59
2010    58
2011    58
2012    58
2013    58
1998    58
2015    57
1997    57
1996    57
1995    56
1994    56
1993    56
1992    56
2014    56
2016    51
2017    51
Name: YEAR, dtype: int64

In [11]:
df.STATE.value_counts()<=14

DISTRICT_OF_COLUMBIA                        False
VIRGINIA                                    False
KANSAS                                      False
NEW_JERSEY                                  False
WASHINGTON                                  False
IOWA                                        False
NEW_YORK                                    False
GEORGIA                                     False
RHODE_ISLAND                                False
LOUISIANA                                   False
IDAHO                                       False
TEXAS                                       False
VERMONT                                     False
CALIFORNIA                                  False
NEBRASKA                                    False
NEVADA                                      False
MISSOURI                                    False
OHIO                                        False
UTAH                                        False
MISSISSIPPI                                 False


In [12]:
df.head()

PRIMARY_KEY       STATE  YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0     1992_ALABAMA     ALABAMA  1992     NaN      2678885.0         304177.0   
1      1992_ALASKA      ALASKA  1992     NaN      1049591.0         106780.0   
2     1992_ARIZONA     ARIZONA  1992     NaN      3258079.0         297888.0   
3    1992_ARKANSAS    ARKANSAS  1992     NaN      1711959.0         178571.0   
4  1992_CALIFORNIA  CALIFORNIA  1992     NaN     26260025.0        2072470.0   

   STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0      1659028.0       715680.0          2653798.0                1481703.0   
1       720711.0       222100.0           972488.0                 498362.0   
2      1369815.0      1590376.0          3401580.0                1435908.0   
3       958785.0       574603.0          1743022.0                 964323.0   
4     16546514.0      7641041.0         27138832.0               14358922.0   

          ...           GRADES_4_G  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  \
0         ...              57948.0     58025.0      41167.0      471564.0   
1         ...               9748.0      8789.0       6714.0       79117.0   
2         ...              55433.0     49081.0      37410.0      437127.0   
3         ...              34632.0     36011.0      27651.0      281338.0   
4         ...             418418.0    363296.0     270675.0     3286034.0   

   GRADES_9_12_G  GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  \
0       196386.0      676174.0        208.327876        252.187522   
1        30847.0      112335.0               NaN               NaN   
2       175210.0      614881.0        215.253932        265.366278   
3       123113.0      405259.0        210.206028        256.312090   
4      1372011.0     4717112.0        208.398961        260.892247   

   AVG_READING_4_SCORE  AVG_READING_8_SCORE  
0           207.963517                  NaN  
1                  NaN           258.859712  
2           206.212716           262.169895  
3           208.634458           264.619665  
4           196.764414                  NaN  

[5 rows x 25 columns]

In [ ]:
df.sort_values(['YEAR','TOTAL_REVENUE'])

In [ ]:
df['STATE'].sort_values()

3. Now, take into account the time factor. Replicate your second answer, but this time, fill in the missing values by using a statistic that is calculated within the year of the observation. For example, if you want to fill in a missing value for a variable with the mean of that variable, calculate the mean by using only the observations for that specific year.

ANSWER: For ENROLL, we can impute a mean by year except for the first and last year which are missing all values for ENROLL. We can backfill 1992 and forward fill 2017.  

The revenue columns and the expenditure columns are missing mostly for certain areas in the US so we should remove areas in the US that contain mostly missing values (value counts < 15 because imputing a value would introduce too much error. Remaining nulls in these columns can be imputed with the median since we're dealing with money. 

The average columns are missing every other year so we should interpolate the mean of the year before and after. 

In [14]:
df['ENROLL'] = df['ENROLL'].fillna(method='ffill')

In [15]:
df['ENROLL'] = df['ENROLL'].fillna(method='bfill')

In [27]:
df.isnull().sum()

PRIMARY_KEY                       0
STATE                             0
YEAR                              0
ENROLL                            0
TOTAL_REVENUE                   212
FEDERAL_REVENUE                 212
STATE_REVENUE                   212
LOCAL_REVENUE                   212
TOTAL_EXPENDITURE               212
INSTRUCTION_EXPENDITURE         212
SUPPORT_SERVICES_EXPENDITURE    212
OTHER_EXPENDITURE               263
CAPITAL_OUTLAY_EXPENDITURE      212
GRADES_PK_G                     173
GRADES_KG_G                     132
GRADES_4_G                      131
GRADES_8_G                      131
GRADES_12_G                     131
GRADES_1_8_G                    131
GRADES_9_12_G                   131
GRADES_ALL_G                    173
AVG_MATH_4_SCORE                956
AVG_MATH_8_SCORE                960
AVG_READING_4_SCORE             959
AVG_READING_8_SCORE             994
dtype: int64

In [ ]:
df.groupby(['YEAR','STATE'])['TOTAL_REVENUE'].median()

In [64]:
# Dropping states with <= 14 value counts
# dropped_states = ['BUREAU_OF_INDIAN_AFFAIRS','BI','DD','NORTHERN_MARIANA_ISLANDS','COMMONWEALTH_OF_MARIANAS','DOD_-_DOMESTIC','DOD_-_OVERSEAS','BUREAU_OF_INDIAN_EDUCATION','BIE','DEPARTMENT_OF_DEFENSE_EDUCATION_ACTIVITY','DOD','DEPARTMENT_OF_DEFENSE','GU','DOD_-_FOREIGN','BUREAU_OF_INDIAN_EDUCATIO','VI','PR','DOD_(OVERSEAS_AND_DOMESTIC_COMBINED)','MP','U.S._VIRGIN_ISLANDS' 'AS','MARIANAS']

df.drop(df[df.STATE == 'DOD_OVERSEAS'].index, inplace=True)

In [65]:
df.STATE.value_counts()

DISTRICT_OF_COLUMBIA    29
VIRGINIA                28
KANSAS                  26
NEW_YORK                26
LOUISIANA               26
IDAHO                   26
TEXAS                   26
VERMONT                 26
CALIFORNIA              26
NEBRASKA                26
NEVADA                  26
NEW_JERSEY              26
MISSOURI                26
UTAH                    26
MISSISSIPPI             26
NEW_HAMPSHIRE           26
OREGON                  26
MAINE                   26
NORTH_DAKOTA            26
ARKANSAS                26
HAWAII                  26
NORTH_CAROLINA          26
TENNESSEE               26
MONTANA                 26
MINNESOTA               26
RHODE_ISLAND            26
GEORGIA                 26
IOWA                    26
DELAWARE                26
INDIANA                 26
WYOMING                 26
COLORADO                26
KENTUCKY                26
SOUTH_CAROLINA          26
OKLAHOMA                26
ARIZONA                 26
MICHIGAN                26
W

In [ ]:
# percentages of missing values
df.isnull().sum()*100/df.isnull().count()

In [67]:
df.isnull().sum()

PRIMARY_KEY                       0
STATE                             0
YEAR                              0
ENROLL                            0
TOTAL_REVENUE                   141
FEDERAL_REVENUE                 141
STATE_REVENUE                   141
LOCAL_REVENUE                   141
TOTAL_EXPENDITURE               141
INSTRUCTION_EXPENDITURE         141
SUPPORT_SERVICES_EXPENDITURE    141
OTHER_EXPENDITURE               192
CAPITAL_OUTLAY_EXPENDITURE      141
GRADES_PK_G                     148
GRADES_KG_G                     116
GRADES_4_G                      116
GRADES_8_G                      116
GRADES_12_G                     116
GRADES_1_8_G                    116
GRADES_9_12_G                   116
GRADES_ALL_G                    148
AVG_MATH_4_SCORE                885
AVG_MATH_8_SCORE                889
AVG_READING_4_SCORE             888
AVG_READING_8_SCORE             923
dtype: int64

In [66]:
# finding percentages of missing values
df.isnull().sum()*100/df.isnull().count()

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                           0.000000
TOTAL_REVENUE                    9.922590
FEDERAL_REVENUE                  9.922590
STATE_REVENUE                    9.922590
LOCAL_REVENUE                    9.922590
TOTAL_EXPENDITURE                9.922590
INSTRUCTION_EXPENDITURE          9.922590
SUPPORT_SERVICES_EXPENDITURE     9.922590
OTHER_EXPENDITURE               13.511612
CAPITAL_OUTLAY_EXPENDITURE       9.922590
GRADES_PK_G                     10.415201
GRADES_KG_G                      8.163265
GRADES_4_G                       8.163265
GRADES_8_G                       8.163265
GRADES_12_G                      8.163265
GRADES_1_8_G                     8.163265
GRADES_9_12_G                    8.163265
GRADES_ALL_G                    10.415201
AVG_MATH_4_SCORE                62.280084
AVG_MATH_8_SCORE                62.561576
AVG_READING_4_SCORE             62

In [ ]:
df.groupby('YEAR').agg({'TOTAL_EXPENDITURE':'mean'})

In [ ]:
df.groupby(['YEAR','STATE'])['TOTAL_REVENUE'].median()

In [ ]:
# checking in any rows have all values missing
df.isnull().sum()

In [ ]:
df.shape

In [ ]:
# dropping any remaining minor null values
# df.dropna(how='all').shape

In [ ]:
df.columns

In [ ]:
# replacing missing values in 'Total_Revenue' with mean of that year
#df.groupby('YEAR')['TOTAL_REVENUE'].apply(lambda x: x.fillna(x.mean()))
#df["Subscribers"].fillna(df["Subscribers"].mean(), inplace=True)
#df.fillna(df.groupby([df.YEAR]).transform('mean'), inplace=True)
df['TOTAL_REVENUE'] = df.groupby('YEAR').TOTAL_REVENUE.transform(lambda x: x.fillna(x.median()))

In [ ]:
df.groupby('YEAR').agg({'TOTAL_REVENUE': 'mean'})

In [ ]:
df.groupby('YEAR').agg({'FEDERAL_REVENUE': 'mean'})

In [ ]:
df.groupby('YEAR').agg({'FEDERAL_REVENUE': 'mean'})

In [ ]:
df.groupby('YEAR').agg({'FEDERAL_REVENUE': 'mean'})

In [ ]:
df.groupby('YEAR').agg({'ENROLL': 'mean'})

In [ ]:
df.groupby('YEAR').agg({'TOTAL_REVENUE': 'mean'})

In [ ]:
df.groupby('YEAR').agg({'FEDERAL_REVENUE': 'mean'})

In [ ]:
df.groupby('YEAR').agg({'STATE_REVENUE': 'mean'})

In [ ]:
df.groupby('YEAR').agg({'LOCAL_REVENUE': 'mean'})

In [ ]:
df.YEAR.plot(kind='hist')
plt.show()